hop/io.py

```diff
@@ -3,44 +3,218 @@
+
+        Raises:
+            ValueError: If the mode is not set to read/write or if more than
+                one topic is specified in write mode.
+
+        """
+        group_id, broker_addresses, topics = kafka.parse_kafka_url(url)
+        logger.debug("connecting to addresses=%s  group_id=%s  topics=%s",
+                     broker_addresses, group_id, topics)
+        if not auth and self.auth:
+            auth = self.auth
+        if mode == "w":
+            if len(topics) != 1:
+                raise ValueError("must specify exactly one topic in write mode")
+            if group_id is not None:
+                warnings.warn("group ID has no effect when opening a stream in write mode")
+            if start_at is not None:
+                warnings.warn("start_at has no effect when opening a stream in write mode")
+            if persist is not None:
+                warnings.warn("read_forever has no effect when opening a stream in write mode")
+            return _open_producer(broker_addresses, topics[0], auth=auth)
+        elif mode == "r":
+            if group_id is None:
+                group_id = _generate_group_id(10)
+                logger.info("group ID not specified, generating a random group ID")
+            # set up extra options if provided
+            opts = {}
+            if start_at or self.start_at:
+                opts["start_at"] = start_at if start_at else self.start_at
+            if persist is not None or self.persist is not None:
+                opts["read_forever"] = persist if persist is not None else self.persist
+            return _open_consumer(
+                group_id,
+                broker_addresses,
+                topics,
+                auth=auth,
+                metadata=metadata,
+                **opts,
+            )
+        else:
+            raise ValueError("mode must be either 'w' or 'r'")
+
+
+class Deserializer(Enum):
+    CIRCULAR = models.GCNCircular
+    VOEVENT = models.VOEvent
+    BLOB = models.MessageBlob
+
+    @classmethod
+    def deserialize(cls, message):
+        """Deserialize a stream message and instantiate a model.
+
+        Args:
+            message: A serialized message.
+
+        Returns:
+            A data container corresponding to the format in the
+            serialized message.
+
+        Raises:
+            ValueError: If the message is incorrectly formatted or
+                if the message format is not recognized.
 
         """
-        opts = {**self._options, **kwargs}
-        return streaming.open(*args, **opts)
+        try:
+            format = message["format"].upper()
+            content = message["content"]
+        except (KeyError, TypeError):
+            raise ValueError("Message is incorrectly formatted")
+        else:
+            if format in cls.__members__:
+                return cls[format.upper()].value(**content)
+            else:
+                raise ValueError(f"Message format {format} not recognized")
```

Comment:
I think that's a good idea. What do you imagine we should expose to the `MessageBlob` to make this clear? Or do you think a log message (maybe at log level warning) would be more suitable?

Reply:
I think a logged warning is definitely a good approach, since it will lead users to actually report the issue. I was thinking just some sort of `blob.missing_schema = True` would be fine for identifying it programmatically.

Reply:
Added this in just now, along with supporting arbitrary JSON-serializable unstructured messages. An extra conditional was added to handle the unstructured case differently from the rest (the previous implementation only supported dictionaries unintentionally). In the case the message format is not recognized, a `MessageBlob` is returned with `missing_schema = True`.